From 1dd86d835b27ed149b66aee1a4853d1db8a1f44c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D0=90=D0=BD=D0=B4=D1=80=D0=B5=D0=B9=20=D0=9A=D0=BE=D0=B2?=
 =?UTF-8?q?=D0=B0=D0=BB=D1=91=D0=B2?=
 <115738313+andr3y-k0v4l3v@users.noreply.github.com>
Date: Thu, 10 Jul 2025 22:24:53 +0300
Subject: [PATCH 12/33] image-zoom.c: add the xsize and ysize check for zero in
 _cfImageZoomNew() (#86)

When calculating z->xmod, z->xstep, etc., if xsize and ysize are zero, division
by zero occurs. To avoid this, the xsize and ysize check for zero is added at
the beginning of the function.

Co-authored-by: Andrey Kovalev <ded@altlinux.org>
---
 cupsfilters/image-zoom.c    | 5 +++--
 cupsfilters/imagetoraster.c | 2 ++
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/cupsfilters/image-zoom.c b/cupsfilters/image-zoom.c
index a614522a..f3c1c5e0 100644
--- a/cupsfilters/image-zoom.c
+++ b/cupsfilters/image-zoom.c
@@ -92,8 +92,9 @@ _cfImageZoomNew(
   if (xsize > CF_IMAGE_MAX_WIDTH ||
       ysize > CF_IMAGE_MAX_HEIGHT ||
       (xc1 - xc0) > CF_IMAGE_MAX_WIDTH ||
-      (yc1 - yc0) > CF_IMAGE_MAX_HEIGHT)
-    return (NULL);		// Protect against integer overflow
+      (yc1 - yc0) > CF_IMAGE_MAX_HEIGHT ||
+      xsize == 0 || ysize == 0)
+    return (NULL);		// Protect against integer overflow and divide by zero
 
   if ((z = (cf_izoom_t *)calloc(1, sizeof(cf_izoom_t))) == NULL)
     return (NULL);
diff --git a/cupsfilters/imagetoraster.c b/cupsfilters/imagetoraster.c
index 950c98de..feb2e790 100644
--- a/cupsfilters/imagetoraster.c
+++ b/cupsfilters/imagetoraster.c
@@ -1680,6 +1680,8 @@ cfFilterImageToRaster(int inputfd,         // I - File descriptor input stream
 			      (doc.Orientation > 1 ? -1 : 1) * xtemp,
 			      (doc.Orientation > 1 ? -1 : 1) * ytemp,
 			      doc.Orientation & 1, zoom_type);
+	  if (z == NULL) continue;
+
 	  //
 	  // Write leading blank space as needed...
 	  //
-- 
2.53.0

